cmd/apply.go

```diff
@@ -32,61 +33,46 @@ var applyCmd = &cobra.Command{
 	Short:   "apply rules from file",
 	Long:    `apply rules described in a configuration file`,
 	Example: c.ProgramName + " apply --file rule.yml",
-	Run: func(cmd *cobra.Command, args []string) {
+	RunE: func(cmd *cobra.Command, args []string) error {
 		rulesContent, err := os.Open(c.RulesFile)
 		if err != nil {
-			fmt.Printf("error opening file: %v", err)
-			os.Exit(1)
+			return fmt.Errorf("opening file: %v", err)
 		}
 		ruleSet, err := rules.NewRuleSetFromFile(rulesContent)
 		if err != nil {
-			fmt.Printf("unable to open rules file: %v\n", err)
-			os.Exit(1)
+			return fmt.Errorf("unable to open rules file: %v", err)
 		}
 		ipt, err := iptables.NewIPTablesInstance()
 		if err != nil {
-			fmt.Printf("unable to get iptables instance: %v\n", err)
-			os.Exit(1)
+			return fmt.Errorf("unable to get iptables instance: %v", err)
 		}
 
-		var wg sync.WaitGroup
-		chErr := make(chan error, len(ruleSet.Rules))
-		chLimit := make(chan int, 10)
+		g := new(errgroup.Group)
 		rulesFileIsValid := true
 
 		for _, rule := range ruleSet.Rules {
-			wg.Add(1)
-			// add slot to buffered channel
-			chLimit <- 1
-			go func(rule iptables.Rule, wg *sync.WaitGroup, chErr chan error, chLimit chan int) {
-				err := ipt.ValidateForward(&rule)
-				wg.Done()
-				chErr <- err
-				// free slot from buffered channel
-				<-chLimit
-			}(rule, &wg, chErr, chLimit)
+			r := &rule
```

Comment:
can you explain why you want to change this?

Reply:
TLDR; when running a goroutine with function closure (`rule` variable) there is for each of them the last element of `ruleSet.Rules` will be used.
Creating a copy (`r := &rule`) ensures that each goroutine processes unique elements.
https://go.dev/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables

Reply:
Ok, got it. That's fine for me, just add a `g.SetLimit(10)` before the `g.Go()` function to limit the number of active goroutines as was previously implemented with `WaitGroup`.